Improve SVG Markers

rwlbuis opened the pull request:
#### 25a983afbde5c58bd75b054a5fdd511e28216d8e
<pre>
Improve SVG Markers
<a href="https://bugs.webkit.org/show_bug.cgi?id=269789">https://bugs.webkit.org/show_bug.cgi?id=269789</a>

Reviewed by Nikolas Zimmermann.

Fix the following problems:
- if the previous segment was a move to, we need to calculate the previous (mid-marker) segment orientation as a start marker
- if we are now moving to a new subpath, the previous (mid-marker) segment orientation should be calculated as an end marker
- don&apos;t calculate the out slope if we are moving to a new suboath

Fixes:
web-platform-tests/svg/painting/reftests/marker-path-011.svg
web-platform-tests/svg/painting/reftests/marker-path-012.svg

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/svg/SVGMarkerData.h:
(WebCore::SVGMarkerData::SVGMarkerData):
(WebCore::SVGMarkerData::updateFromPathElement):

Canonical link: <a href="https://commits.webkit.org/275167@main">https://commits.webkit.org/275167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0b4b42adaa0881053a9f13d76dfab57296f6cea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41071 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43449 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43632 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37163 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23106 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17415 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41645 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17002 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35371 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44945 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37257 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36683 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13021 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38809 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17505 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9217 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17149 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGMarkerData.h:
```diff
@@ -144,12 +152,13 @@ class SVGMarkerData {
     FloatPoint m_subpathStart;
     FloatPoint m_inslopePoints[2];
     FloatPoint m_outslopePoints[2];
     bool m_reverseStart;
```
You could also init this variable here inline.

rwlbuis replied:
Well, it is assigned a value in the (only) ctor :)

rwlbuis committed:
Improve SVG Markers
https://bugs.webkit.org/show_bug.cgi?id=269789

Reviewed by Nikolas Zimmermann.

Fix the following problems:
- if the previous segment was a move to, we need to calculate the previous (mid-marker) segment orientation as a start marker
- if we are now moving to a new subpath, the previous (mid-marker) segment orientation should be calculated as an end marker
- don't calculate the out slope if we are moving to a new suboath

Fixes:
web-platform-tests/svg/painting/reftests/marker-path-011.svg
web-platform-tests/svg/painting/reftests/marker-path-012.svg

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/svg/SVGMarkerData.h:
(WebCore::SVGMarkerData::SVGMarkerData):
(WebCore::SVGMarkerData::updateFromPathElement):

Canonical link: https://commits.webkit.org/275167@main
